File: frontend/packages/shared-normalizers/src/group.ts
import type {
  Group,
  GroupMember,
  RawGroupDTO,
  RawGroupMemberDTO,
} from "@im/shared-types";
import {
  asNumber,
  asString,
  isRawGroup,
  isRawGroupMember,
  isRecord,
} from "@im/shared-types";

const normalizeMemberRole = (role: unknown): GroupMember["role"] => {
  if (role === 3 || role === "3" || role === "OWNER") return "OWNER";
  if (role === 2 || role === "2" || role === "ADMIN") return "ADMIN";
  return "MEMBER";
};

export const normalizeGroup = (raw: RawGroupDTO | Group | unknown): Group => {
  const record = isRawGroup(raw) ? raw : {};
  const looseRecord = isRecord(raw) ? raw : {};
  const groupName = asString(record.groupName ?? looseRecord.group_name ?? record.name);
  return {
    id: asString(record.id ?? looseRecord.groupId ?? looseRecord.group_id),
    name: asString(record.name) || undefined,
    groupName,
    description:
      asString(record.description ?? record.announcement) || undefined,
    announcement: asString(record.announcement) || undefined,
    type: record.type,
    avatar: asString(record.avatar) || undefined,
    ownerId: asString(record.ownerId ?? looseRecord.owner_id),
    memberCount: asNumber(record.memberCount ?? looseRecord.member_count, 0),
    maxMembers: Number.isFinite(asNumber(record.maxMembers, Number.NaN))
      ? asNumber(record.maxMembers)
      : undefined,
    status: record.status,
    unreadCount: Number.isFinite(asNumber(record.unreadCount, Number.NaN))
      ? asNumber(record.unreadCount)
      : undefined,
    lastMessageTime: asString(record.lastMessageTime) || undefined,
    lastActivityAt: asString(record.lastActivityAt) || undefined,
    createTime: asString(record.createTime),
  };
};

export const normalizeGroupMember = (
  raw: RawGroupMemberDTO | GroupMember | unknown,
): GroupMember => {
  const record: RawGroupMemberDTO = isRawGroupMember(raw) ? raw : {};
  const looseRecord = isRecord(raw) ? raw : {};
  return {
    id: asString(record.id) || undefined,
    groupId: asString(record.groupId ?? looseRecord.group_id) || undefined,
    userId: asString(record.userId ?? looseRecord.user_id ?? record.id),
    username: asString(record.username) || undefined,
    nickname: asString(record.nickname) || undefined,
    avatar: asString(record.avatar) || undefined,
    role: normalizeMemberRole(record.role),
    joinTime: asString(record.joinTime),
  };
};

File: flutter/packages/core/lib/src/models/group.dart
// ignore_for_file: invalid_annotation_target

import 'package:freezed_annotation/freezed_annotation.dart';

part 'group.freezed.dart';
part 'group.g.dart';

@freezed
class Group with _$Group {
  const factory Group({
    required String id,
    required String name,
    String? avatar,
    String? description,
    String? ownerId,
    int? memberCount,
    String? createTime,
    String? updateTime,
  }) = _Group;

  factory Group.fromJson(Map<String, dynamic> json) => _$GroupFromJson(json);
}

String? _roleFromJson(dynamic value) => value?.toString();

@freezed
class GroupMember with _$GroupMember {
  const factory GroupMember({
    required String id,
    required String userId,
    required String groupId,
    String? nickname,
    @JsonKey(fromJson: _roleFromJson) String? role,
    String? joinTime,
  }) = _GroupMember;

  factory GroupMember.fromJson(Map<String, dynamic> json) =>
      _$GroupMemberFromJson(json);
}
